# **Numpy**


###**What is Numpy?**

=> Numpy is a Python library or fundamental package, which is used for working with arrays. We can perform many scientific computiation on those arrays and also it is 5-100 times faster than traditional Python lists. Numpy is short for Numerical Python.

Numpy in partially written in python but for fast calculation most of its codes are written in C or C++.

Numpy provides
  1. An array object of arbitrary homogeneous items
  2. Fast mathematical operations over arrays
  3. Linear Algebra, Fourier Transforms, Random Number Generation


---



### **How to create basic arrays using Numpy?**

=>First we need to import the Numpy library to work with it. We can create Numpy ndarray object using array() function.

In [ ]:
#Creating a Numpy array

#Import the Library
import numpy as np

arr = np.array([1,2,3]) 

# This will print single dimensional array
print(arr)

[1 2 3]


We can pass list, tuple or array-like object in the array() method to create an ndarray.

In [ ]:
import numpy as np

tup=(5,6,7,8)                  # Tuple
arr_tup=np.array(tup)          # Passing tuple in array function

print(arr_tup)                 # This will print 1-D array

[5 6 7 8]


We  can create array filled with zeros or ones.

In [ ]:
import numpy as np

arr_0=np.zeros(4) # Using np.zeros() we can create an array of zeros. Here 4 is the number of elements (zeros)
arr_1=np.ones(7)  # Using np.ones() we can create an array of ones. Here 7 is the number of elements (ones)

print(arr_0)
print(arr_1)

[0. 0. 0. 0.]
[1. 1. 1. 1. 1. 1. 1.]


We can create an array with a range of elements with the help of np.arange() function. We can also specify the first number, last number and step size.

In [ ]:
import numpy as np

arr_1=np.arange(7)      # This will create an array starting from 0 to 6
arr_2=np.arange(2,9,2)  # This will create an array [2,4,6,8] as the starting point=2, ending point=9, step size=2

print(arr_1)
print(arr_2)

[0 1 2 3 4 5 6]
[2 4 6 8]


We can also use np.linspace() to create an array with values that are spaced linearly in a specified interval.

In [ ]:
import numpy as np
arr=np.linspace(0, 10, num=5)     # This will create an array between the range 0 to 10 and there is 65 elements in this range.



---




###**Sorting and Concatenation of Array Elements**

Sorting :

We can sort elements of numpy array using np.sort() function.

In [ ]:
import numpy as np

arr=np.array([8,4,9,5,1,0,6,2,3,7])  # Unsorted array

print(arr)                           # Printing Unsorted Array

arr_sorted = np.sort(arr)            # Array Sorting

print(arr_sorted)                    # Printing Sorted Array

[8 4 9 5 1 0 6 2 3 7]
[0 1 2 3 4 5 6 7 8 9]


Concatenation :

We can concatenate arrays using the np.concatenate() function.

In [ ]:
import numpy as np

a = np.array([1, 2, 3, 4])
b = np.array([5, 6, 7, 8])

c=np.concatenate((a, b))

print(c)

[1 2 3 4 5 6 7 8]




---



### **Dimensions of Arrays**

Arrays can be of different dimensions.

0-Dimensional Array:

In [ ]:
import numpy as np

arr = np.array(10)

print(arr)

10


1-Dimensional Array:

In [ ]:
import numpy as np

arr = np.array([1, 2, 3, 4, 5])

print(arr)

[1 2 3 4 5]


2-Dimensional Array:

In [ ]:
import numpy as np

arr = np.array([[1, 2, 3], [4, 5, 6]])    # We need to pass two lists separated by a comma to create a 2-D array

print(arr)

[[1 2 3]
 [4 5 6]]


3-Dimensional Array:

In [ ]:
import numpy as np

arr = np.array([[[1, 2, 3], [4, 5, 6]], [[1, 2, 3], [4, 5, 6]]])

print(arr)

[[[1 2 3]
  [4 5 6]]

 [[1 2 3]
  [4 5 6]]]


Higher Dimensional Array:

We can create higher dimensional arrays by spcifying number of dimensions using ndmin arguement.

In [ ]:
import numpy as np

arr = np.array([1, 2, 3, 4], ndmin=5)         # Defining array dimension

print(arr)                                    #Printing the array

print('Number of dimensions :', arr.ndim)     # Printing array dimensions

[[[[[1 2 3 4]]]]]
Number of dimensions : 5




---



### **Array Size and Shape**

ndarray.size will tell us the total number of elements of the array. This is the product of the elements of the array’s shape.

ndarray.shape will display a tuple of integers that indicate the number of elements stored along each dimension of the array.

In [ ]:
import numpy as np

arr=np.array([[1,2,3],[4,5,6]])

print(arr)
print("Size of the array : ",arr.size)
print("Shape of the array : ",arr.shape)

[[1 2 3]
 [4 5 6]]
Size of the array :  6
Shape of the array :  (2, 3)




---



###**Array Reshape**

Using arr.reshape() will give a new shape to an array without changing the data. We need to remember that when we are using the reshape method, the array we want to produce needs to have the same number of elements as the original array.

In [ ]:
import numpy as np

a=np.array([1,2,3,4,5,6]) # 1-D array

print(a)

b=a.reshape(2,3) # 2-D Array. The command will be like - arr_name.reshape(no. of rows, no. of columns)

print(b)

[1 2 3 4 5 6]
[[1 2 3]
 [4 5 6]]




---



### **Array Indexing**

We can access the elements of numpy array through indexes.

For 1-D array:

In [ ]:
#Accessing 1-D array elements using indexes
 
import numpy as np

arr = np.array([11, 22, 33, 44])

#Printing the array element at index 2
print(arr[2])

33


For 2-D array:

We can use comma separated integers representing the dimension and the index of the element.We can use this style of array indexing in higher order array indexing. For instance, in the case of 3-D array it will be like arr[0,1,2].

In [ ]:
# Accessing 2-D array elements using indexes

import numpy as np

arr = np.array([[1,2,3,4,5], [6,7,8,9,10]])

print('4th element on 1st dim: ', arr[0, 3]) 
print('3th element on 2nd dim: ', arr[1, 2]) 

4th element on 1st dim:  4
3th element on 2nd dim:  8


Negative Indexing:

In [ ]:
import numpy as np

arr = np.array([[1,2,3,4,5], [6,7,8,9,10]])

print('Last element from 2nd dim: ', arr[1, -1])      # Accessing last element of the 2nd dim



---



### **Array Slicing**

We can also slice ndarray. It is similiar to regular slicing.

Syntax of slicing array is-> arr [ start : end : step(optional) ]

In [ ]:
# Array slicing

import numpy as np

arr = np.array([1, 2, 3, 4, 5, 6, 7])

print(arr[1:5])                          # arr[1:5] means we are starting from index 1 and we will go upto index 4



[2 3 4 5]




---



### **Numpy Datatypes**

Numpy has a lots of datatypes. The list is given below-

i -> integer

b -> boolean


u -> unsigned integer


f -> float


c -> complex float


m -> timedelta


M -> datetime


O -> object


S -> string


U -> unicode string


V -> fixed chunk of memory for other type ( void )

### **Checking of Array Datatype**:

We can know the datatype of the array using the property dtype.

In [ ]:
import numpy as np

arr = np.array(['apple', 'banana', 'cherry'])

print(arr.dtype)

<U6


### **Creating Arrays With a Defined Data Type**

We use the array() function to create arrays, this function can take an optional argument: dtype that allows us to define the expected data type of the array elements.

In [ ]:
import numpy as np

arr = np.array([1, 2, 3, 4], dtype='S')         # Datatype is defined as string

print(arr)
print(arr.dtype)

[b'1' b'2' b'3' b'4']
|S1


For i, u, f, S, and U we can define size as well.

In [ ]:
import numpy as np

arr = np.array([1, 2, 3, 4], dtype='i4')        # array with data type 4 bytes integer

print(arr)
print(arr.dtype)

[1 2 3 4]
int32


### **Converting Data Type on Existing Arrays**

By using astype() function we can create a copy of the array, and allows us to specify the data type as a parameter. For example, we can define the datatype integer by writing int or 'i', and for floating point numbers float or 'f'.

In [ ]:
import numpy as np

arr = np.array([1.1, 2.1, 3.1])      # Array consisting floating point numbers

newarr = arr.astype(int)             # new array created which is the copy of arr but having integer values of the corresponding elements in arr

print(newarr)
print(newarr.dtype)

[1 2 3]
int64


In [ ]:
import numpy as np

arr = np.array([1, 0, 3, 0, 999])

newarr = arr.astype(bool)               # we can also convert integer to boolean

print(newarr)
print(newarr.dtype)

[ True False  True False  True]
bool




---



### **Comparing Speed of Numpy Arrays and Lists**

Numpy arrays are 5-100 times faster than python lists.

In [ ]:
import time                               # Time module

py_list=[i for i in range(10000000)]      # Python list --> [0,1,2,3,.....,9999999]

start=time.time()                         # starting time
py_list=[i+100 for i in py_list]          # iterating over python list through a for loop and adding 100 with each element
end=time.time()                           # ending time

elapsed_time=end-start                    # total time taken to complete the work (for loop iteration)

print(round(elapsed_time,6))              # printing the time elapsed upto 6 digit precision

0.641966


In [ ]:
import time
import numpy as np

np_arr=np.array([i for i in range(10000000)])     # numpy array

start=time.time()
np_arr+=100                                       # adding 100 with each element in rthe array
end=time.time()

elapsed_time=end-start

print(round(elapsed_time,6))

0.006083


From the above experiment you can see that numpy is 105.534 times (0.641966 / 0.006083) faster than python list. 

The reason for Numpy arrays being faster is -

* Numpy array is a collection of similar data-types that are densely packed in memory. A Python list can have different data-types, which puts lots of extra constraints while doing computation on it.
* Numpy is able to divide a task into multiple subtasks and process them parallelly.
* Numpy functions are implemented in C. Which again makes it faster compared to Python Lists.





---





---



### **Numpy Basic Operations**

*   Addition, Subtraction, Multiplication, Division :

To add two ndarrays we need to use + operator. Similarly for subtraction, multiplication, division we will use -, *, / operator respectively.

In [ ]:
import numpy as np

a=np.array([1,2,3])
b=np.array([4,5,6])

add=a+b     # Addition
sub=a-b     #Subtraction
mul=a*b     #Multiplication
div=a/b     #Division

print("Addition matrix : ",add)
print("Subtraction matrix : ",sub)
print("Multiplication matrix : ",mul)
print("Division matrix : ",div)

Addition matrix :  [5 7 9]
Subtraction matrix :  [-3 -3 -3]
Multiplication matrix :  [ 4 10 18]
Division matrix :  [0.25 0.4  0.5 ]


* Sum:


We can get the summation of the array elements using sum() function.

In [ ]:
import numpy as np

arr=np.array([ 1, 2, 3, 4, 5])

summ = arr.sum()                  # Storing the summation of the array elements

print(summ)

15


We can sum over the axis of rows with axis=0 and over columns with axis=1.

In [ ]:
import numpy as np

arr = np.array([[1,2],[3,4]])

# Printing the 2-D array
print(arr)

# Printing the sum over the axis of rows
print(arr.sum(axis=0))

# Printing the sum over the axis of columns
print(arr.sum(axis=1))

[[1 2]
 [3 4]]
[4 6]
[3 7]


* Maximum and Minimum:

For evaluating maximum and minimum we use max() and min() functions.

In [ ]:
import numpy as np

arr = np.array([1,2,654,12,35,0,-302])

# Minimum
minimum=arr.min()
print("Minimum : ",minimum)

# Maximum
maximum=arr.max()
print("Maximum : ",maximum)

Minimum :  -302
Maximum :  654


* Exponential and Logarithmic operation:

we can use exp() and log() function for exponential and logarithmic operation.

In [ ]:
import numpy as np

arr=np.array([1, 2, 3, 4, 5])

# Exponent
b= np.exp(arr)
print("Exponent : ",b)

#Logarithm
c=np.log(arr)
print("Logarithm : ",c)

Exponent :  [  2.71828183   7.3890561   20.08553692  54.59815003 148.4131591 ]
Logarithm :  [0.         0.69314718 1.09861229 1.38629436 1.60943791]




---



### **List Comprehension**

List comprehensions are used for creating new lists from other iterables like tuples, strings, arrays, lists, etc.

Syntax :

ListName = [ expression(element) for element in oldList if condition ]

Advantages:

* More time efficient and space efficient than loops.
* Require fewer lines of code.
* This is faster method.

In [2]:
#List Comprehension
List_1=[i for i in range(20)]
List_2=[j for j in "Hello World!"]
List_3=[k for k in (15,35,60,26,37,18)]

print(List_1)
print(List_2)
print(List_3)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
['H', 'e', 'l', 'l', 'o', ' ', 'W', 'o', 'r', 'l', 'd', '!']
[15, 35, 60, 26, 37, 18]




---

